src/ast/traits.rs

```diff
@@ -0,0 +1,290 @@
+pub trait AsNode {
+    fn as_node(&self) -> &dyn Node;
+}
+
+impl<T: Node + Sized> AsNode for T {
+    fn as_node(&self) -> &dyn Node {
+        self
+    }
+}
+
+pub(super) type ParseResult<T> = Result<T, MissingEndError>;
+
+/// Only explicitly *implemented* for Keyword nodes.
+/// Only explicitly *called* by the Parse impl of a branch node.
+///
+/// In the event of an error, a branch node's Parse impl will catch the error,
+/// stop parsing the rest of the node (leaving subsequent fields default-initialized),
+/// and begin an unwind (see [ParserStatus::unwinding]).
+/// In effect, this short-circuits the parse, but a JobList is sometimes able
+/// to "stop" the unwind and resume parsing another job.
+///
+/// Either way, the top-level parse operation is infallible in terms of Rust idioms,
+/// with error information stored elsewhere in the [Ast] struct and some nodes being
+/// unsourced after an error.
+pub(super) trait TryParse: Sized {
+    fn try_parse(pop: &mut Populator<'_>) -> ParseResult<Self>;
+}
+
+pub(super) trait Parse: TryParse {
+    fn parse(pop: &mut Populator<'_>) -> Self;
+}
+
+impl<T: Parse> TryParse for T {
+    fn try_parse(pop: &mut Populator<'_>) -> ParseResult<Self> {
+        Ok(Self::parse(pop))
+    }
+}
+
+/// This is for optional values and for lists.
+pub(super) trait CheckParse: TryParse {
+    /// A true return means we should descend into the production, false means stop.
+    fn can_be_parsed(pop: &mut Populator<'_>) -> bool;
+}
+
+impl<T: CheckParse + TryParse> TryParse for Option<T> {
+    fn try_parse(pop: &mut Populator<'_>) -> ParseResult<Self> {
+        if T::can_be_parsed(pop) {
+            Some(pop.try_parse::<T>())
+        } else {
+            None
+        }
+        .transpose()
+    }
+}
+
+/// A list of nodes is parsed by calling CheckParse::can_be_parsed(pop) and T::parse(pop)
+/// repeatedly, until the former returns false.
+impl<T: CheckParse + Parse + ListElement> Parse for Box<[T]> {
+    fn parse(pop: &mut Populator<'_>) -> Self {
+        pop.parse_list(false)
+    }
+}
+
+/// A node that can be parsed as an element of a homogenous list.
+pub(super) trait ListElement: Sized + Node {
+    fn list_kind(list: &[Self]) -> Kind<'_>;
+
+    /// Return whether a list of these nodes allows arbitrary newlines in it.
+    fn chomps_newlines(pop: &Populator<'_>) -> bool;
+    /// Return whether a list of these nodes allows arbitrary semicolons in it.
+    fn chomps_semis(pop: &Populator<'_>) -> bool;
+    /// Return whether a list of these nodes should recover from errors.
+    /// That is, whether we should stop unwinding when we encounter this node's parent list.
+    fn stops_unwind(_pop: &Populator<'_>) -> bool {
+        false
+    }
+}
+
+impl<'a, N: ListElement> Node for &'a [N] {
+    fn kind(&self) -> Kind<'_> {
+        N::list_kind(self)
+    }
+}
+
+impl<N: ListElement> Node for Box<[N]> {
+    fn kind(&self) -> Kind<'_> {
+        N::list_kind(self)
+    }
+}
```

Comment:
(Just a brain dump:)
I wonder if we really want every `Box<[N]>` to implement `Node`?
That would make it easy to construct a new list node from existing
nodes, without going through the parser.
I guess that's not a problem in practice.
But if we actually wanted to guard this, we could use the newtype
pattern (similar to what we had before, but make it generic, i.e.

	struct List<N> ( Box<[N]> )